diff --git a/lua/spectre/init.lua b/lua/spectre/init.lua
index 81ef804..db18824 100644
--- a/lua/spectre/init.lua
+++ b/lua/spectre/init.lua
@@ -44,8 +44,8 @@ end
 M.check_replace_cmd_bins = function()
     if state.user_config.default.replace.cmd == 'sed' then
         if vim.loop.os_uname().sysname == 'Darwin' then
-            config.replace_engine.sed.cmd = 'gsed'
-            if vim.fn.executable('gsed') == 0 then
+            config.replace_engine.sed.cmd = 'sed'
+            if vim.fn.executable('sed') == 0 then
                 print("You need to install gnu sed 'brew install gnu-sed'")
             end
         end
